In [1]:
import numpy as np
from scipy.stats import t

## 📌 Question 1 : Paired t-Test (Blood Pressure)

Blood pressure (mmHg) measured before and after medication for 8 patients:  
Before: `[140, 138, 150, 145, 142, 148, 135, 152]`  
After: `[132, 130, 142, 138, 135, 140, 128, 144]`  
At α = 0.05, test whether the medication significantly reduces blood pressure. Also compute a 95% CI for the mean reduction.

#### Step 1 : Hypotheses

Let $\mu_d$ = true mean of (before - after) differences.

$$H_0: \mu_d = 0 \quad \text{(no change)}$$
$$H_1: \mu_d > 0 \quad \text{(blood pressure decreased, one-tailed)}$$

#### Step 2 : Why this test

The same 8 patients are measured twice, so the observations are paired. A paired t-test reduces between-subject noise by working directly on the within-subject differences. With the differences treated as a single sample, this collapses to a one-sample t-test on $d_i = \text{before}_i - \text{after}_i$. Population variance is unknown, so t (not Z) is correct.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{s_d / \sqrt{n}}, \quad df = n - 1$$

where $\bar{d}$ is the mean difference, $s_d$ is the sample standard deviation of the differences.

In [2]:
before = np.array([140, 138, 150, 145, 142, 148, 135, 152])  # before readings
after  = np.array([132, 130, 142, 138, 135, 140, 128, 144])  # after readings
alpha  = 0.05                                                  # significance level
n      = len(before)                                           # sample size

In [3]:
d     = before - after          # paired differences
d_bar = d.mean()                # mean difference
sd    = np.std(d, ddof=1)       # sample std of differences
se    = sd / np.sqrt(n)         # standard error
df    = n - 1                   # degrees of freedom
t_stat = d_bar / se             # test statistic

#### Step 4 : Critical Value Method

In [4]:
t_crit = t.ppf(1 - alpha, df)          # one-tailed critical value
print(f't_stat : {t_stat:.3f}')
print(f't_crit : {t_crit:.3f}')

t_stat : 41.671
t_crit : 1.895


**t_stat (41.67) > t_crit (1.895) → Reject H₀**

#### Step 5 : p-value Method

In [5]:
pval = t.sf(t_stat, df)                # one-tailed p-value (upper tail)
print(f'p-val : {pval:.12f}')

p-val : 0.000000000598


**p-val ≈ 0.0000000006 < 0.05 → Reject H₀**

#### Step 6 : Confidence Interval

A 95% CI for $\mu_d$ uses the two-tailed critical value:

$$\bar{d} \pm t_{\alpha/2,\, df} \cdot \frac{s_d}{\sqrt{n}}$$

In [6]:
t_ci  = t.ppf(1 - alpha/2, df)         # two-tailed critical value for CI
moe   = t_ci * se                       # margin of error
lo    = d_bar - moe                     # lower bound
hi    = d_bar + moe                     # upper bound
print(f'95% CI : ({lo:.3f}, {hi:.3f})')

95% CI : (7.192, 8.058)


#### ✅ Final Conclusion

| Metric | Value |
|---|---|
| Mean difference $\bar{d}$ | 7.625 mmHg |
| Std deviation $s_d$ | 0.518 |
| Standard error | 0.183 |
| t statistic | 41.671 |
| t critical (one-tailed) | 1.895 |
| p-value | ≈ 0.0000000006 |
| 95% CI for $\mu_d$ | (7.192, 8.058) |
| Decision | Reject H₀ |

Strong evidence that the medication reduces blood pressure. The mean reduction is about 7.6 mmHg, and the 95% CI (7.19, 8.06) sits entirely above zero, confirming the effect is real and practically meaningful.